### PR TITLE
New Gamemode: King of the Disk

### PR DESCRIPTION
**Pull request by Kierany9**
King of the Disk

King of the Disk is a subtype of the Traitor gamemode that revolves around a huge Syndicate infiltration of the station. However, these traitors aren't here to sabotage NT facilities or eliminate potential double agents, they're here to prove their worth. Equipped with nothing but their wits, a restricted uplink and a pinpointer, the traitors must find the nuclear authentication disk and secure it, whoever manages to bring it back with them to Central Command alive and free will be the winner and be made a full fledged Syndicate Operative. While these rookie traitors only start out with half the telecrystals a normal operative does, the Syndicate is monitoring their every move and are more than willing to lend a hand to the most sucessful operatives. For each (BYOND) minute a traitor spends holding the disk, they gain one extra telecrystal. Whoever ends up with the most telecrystals at the end of the round will also count as a winner, even if they aren't holding the nuke disk!

### Changes from default Traitor
- 45 players are required to start a King of the Disk round.
- King of the Disk has a lot more traitors than a regular Traitor round and slightly more than a Double Agents round.
- All traitors spawn with a pinpointer at the start of the round.
- Traitors have two objectives, Escape Alive and a special 'Steal the Nuke Disk' objective, that can be completed by either having the nuke disk at the end of the round or having the most accumulated TC out of all the surviving traitors on the shuttle/pods.
- Traitors only start with 10 telecrystals in their PDA uplink.
- For every BYOND minute(about 90 real seconds) a traitor spends with the nuke disk and his PDA in his inventory, he gains one extra TC. Mindslaves carrying the disk will automatically transfer the TC gained to their owner.
- The following items are unable to be purchased in King of the Disk: Revolver, Ebow, Surplus Crate and Syndicate Bundle.
- The Cryptographic Sequencer costs 8 TC instead of the standard 6 in King of the Disk mode.
- Spare pinpointers can be purchased for 2 TC. This can be useful if you somehow manage to lose yours or recruit an ally either via mindslave or via golem.
- Syndicate Bundles obtained through cargo have 14 pinpointers inside of them. Enough for all the family!

### CODE REVIEW

**Comment by TheOneAndOnlyCreeperJoe** [comment]
it's ok. +1 i guess.

But hey at least it's original content despite it being Highlander in a different jacket.

**Comment by JohnGinnane** [comment]
Missing a changelog

**Comment by calibraptor** [comment]
Awesome +1

**Comment by GeneralGoose** [comment]
sounds like a funny meme
can't you make king_disk emag a child of emag so you only change it's price?

**Comment by ArcLumin** [comment]
I like this. We need more modes

**Comment by McBawbaggings** [comment]
Won't this turn into a murderbone fest moreso than DA? I mean I get the TC is limited but there are more antags.

**Comment by ArcLumin** [comment]
Yeah, but it also promotes more cautious play than double agents. On top of that, it's even more likely someone in this will go double agent and work with the captain or security, or spending TC's buying mindslaves.

**Comment by McBawbaggings** [comment]
Why would they work together if they are required to backstab each other at some point?

**Comment by ArcLumin** [comment]
They're not requires to backstab eachother. Essentially, by letting one of them hold the disk, they give gear to the Security force, and antags automatically come to them for them to fight. It's a win/win, and they're not required to backstab eachother.

**Comment by McBawbaggings** [comment]
"Traitors have two objectives, Escape Alive and a special 'Steal the Nuke Disk' objective, that can be completed by either having the nuke disk at the end of the round or having the most accumulated TC out of all the surviving traitors on the shuttle/pods."

**Comment by ArcLumin** [comment]
And where does that say that a traitor has to attack/kill/backstab security?

**Comment by McBawbaggings** [comment]
Oh I'm retarded and I see what you mean.

**Comment by Kierany9** [comment]
>tfw misunderstanding

**Comment by ArcLumin** [comment]
Also, because it promotes mindslaves, you could give those mindslaves out to people and have them supplement the security force.

**Comment by Kierany9** [comment]
Added logs since everything is pretty much done

File: code/datums/uplink_item.dm [hunk 1]
```diff
@@ -480,13 +494,13 @@ var/list/uplink_items = list()
 	name = "Tactical Gloves"
 	desc = "Comfortable grey gloves with the CQC martial art inside."
 	item = /obj/item/clothing/gloves/cqc
-	cost = 12
+	cost = 8
```
Comment (Carbonhell) [inline code/datums/uplink_item.dm]: stealth lowering prices,wew

File: code/datums/uplink_item.dm [hunk 2]
```diff
@@ -480,13 +494,13 @@ var/list/uplink_items = list()
...
 
 /datum/uplink_item/dangerous/wrestling
 	name = "Wrestling Belt"
 	desc = "A powerful wrestling moveset given to you via championship belt."
 	item = /obj/item/weapon/storage/belt/champion/wrestling
-	cost = 12
+	cost = 6
```
Comment (Carbonhell) [inline code/datums/uplink_item.dm]: WEW

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 3]
```diff
@@ -410,6 +411,27 @@ This is here to make the tiles around the station mininuke change when it's arme
 		get(src, /mob) << "<span class='danger'>You can't help but feel that you just lost something back there...</span>"
 		qdel(src)
 
+	if(istype(ticker.mode,/datum/game_mode/traitor/king_disk))
+		for(var/obj/item/weapon/disk/nuclear/N in poi_list)
+			var/atom/disk_loc2 = N.loc
```
Comment (Carbonhell) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: wouldn't this code fit better in ticker fire proc or idk,gotta check

File: code/datums/uplink_item.dm (hunk 1 above)
Comment (Carbonhell) [inline code/datums/uplink_item.dm]: i'm a retard but still why this in this pr

**Comment by Lawligagger** [comment]
yes

**Comment by Kokojos** [comment]
Sure let's try it. +1/3rd.

File: code/datums/uplink_item.dm [hunk 4]
```diff
@@ -542,6 +556,7 @@ var/list/uplink_items = list()
 	desc = "A speed loader that contains seven additional .357 Magnum rounds for the syndicate revolver. For when you really need a lot of things dead."
 	item = /obj/item/ammo_box/a357
 	cost = 2
+	excludefrom = list(/datum/game_mode/traitor/king_disk)
```
Comment (Kierany9) [inline code/datums/uplink_item.dm]: Because I thought it'd be a good idea to do some balance changes at the time. Moving to a separate PR.

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 5]
```diff
@@ -410,6 +411,27 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+			if(istype(disk_loc2, /mob))
```
Comment (Carbonhell) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: if(ismob(disk_loc2))

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 6]
```diff
@@ -410,6 +411,27 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+				var/mob/M = disk_loc2
+				var/datum/mind/traitor = M.mind
+				if(traitor.special_role)
```
Comment (Carbonhell) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: who says mob has a mind,might not have it,pls check

File: code/game/gamemodes/nuclear/nuclearbomb.dm [hunk 7]
```diff
@@ -410,6 +411,27 @@ This is here to make the tiles around the station mininuke change when it's arme
...
+					if(king_timer >= 60)
+						king_timer = 0
+						var/list/all_items = traitor.current.GetAllContents()
+						if(traitor.special_role == "Mindslave")
+							for(var/datum/objective/protect/P in traitor.objectives)
+								all_items = P.target.current.GetAllContents()
```
Comment (Carbonhell) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: who says the target's still alive and not gibbed,pls check

File: code/game/gamemodes/objective.dm [hunk 8]
```diff
@@ -521,6 +521,38 @@ var/global/list/possible_items_special = list()
...
+
+
+/datum/objective/steal/disk/check_completion()
+	var/obj/item/device/uplink/O = owner.find_syndicate_uplink()
+	if(!isliving(owner.current))	return 0
+	if(O)
+		var/leader = 1
+		var/turf/location = get_turf(owner.current)
+		for(var/mob/living/player in player_list)
+			if(player.mind.special_role && player.mind != owner)
```
Comment (Carbonhell) [inline code/game/gamemodes/objective.dm]: check if player mind exists just to be sure,again

File: code/game/gamemodes/objective.dm [hunk 9]
```diff
@@ -521,6 +521,38 @@ var/global/list/possible_items_special = list()
...
+				if(player.stat != DEAD)
+					if(istype(player, /mob/living/silicon))
+						continue
+					if(player.onCentcom())
+						if(!istype(get_turf(player.mind.current), /turf/simulated/floor/plasteel/shuttle/red))
```
Comment (Carbonhell) [inline code/game/gamemodes/objective.dm]: player.mind.current?wtf,isn't that just player

File: code/game/gamemodes/traitor/king_disk.dm [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/datum/game_mode/traitor/king_disk
+	name = "king of the disk"
+	config_tag = "king_disk"
+	required_players = 45
+	required_enemies = 8
+	recommended_enemies = 10
+	reroll_friendly = 0
+	restricted_jobs = list("Cyborg", "AI")
```
Comment (Carbonhell) [inline code/game/gamemodes/traitor/king_disk.dm]: does that mean sec can be a king of the disk thing, no right

File: code/game/gamemodes/traitor/king_disk.dm [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+/datum/game_mode/traitor/king_disk/forge_traitor_objectives(datum/mind/traitor)
+	// Steal the disk
+	var/datum/objective/steal/disk/disk_objective = new
+	disk_objective.owner = traitor
+	disk_objective.give_special_equipment()
+	traitor.objectives += disk_objective
+
+	// Escape
+	if(issilicon(traitor.current))
```
Comment (Carbonhell) [inline code/game/gamemodes/traitor/king_disk.dm]: how can borgs even steal the disk...?that doesn't make sense

File: code/game/gamemodes/traitor/king_disk.dm (hunk 10 above)
Comment (Kierany9) [inline code/game/gamemodes/traitor/king_disk.dm]: no

File: code/game/gamemodes/traitor/king_disk.dm (hunk 11 above)
Comment (Kierany9) [inline code/game/gamemodes/traitor/king_disk.dm]: woops

**Comment by carlarc** [comment]
fuck yes.
yes.
this is literally the only good gamemode other than nukeops and it isn't even added yet, we should just run those two

**Comment by Kierany9** [comment]
Will need @nalarmg since I've added stuff to the config for this new mode.

**Comment by Kierany9** [comment]
Should probably mention that at the moment a King of the Disk round has a 3 in 29 chance of occuring, or just over 10%, the same chance as Double Agents, although it'll be less common than a lot of other modes since it needs at least 45 players to even start

**Comment by Badmex** [comment]
yes pls +1/3 also yeah its nice that it needs such a high playerbase to even start, still maybe 40 would be better?

File: code/game/gamemodes/nuclear/nuclearbomb.dm (hunk 3 above)
Comment (Chnkr) [inline code/game/gamemodes/nuclear/nuclearbomb.dm]: This code would fit better in a process.

**Comment by Zadeon** [comment]
yes

**Comment by Nebyab** [comment]
yeahs syre

**Comment by Carbonhell** [comment]
Need to move that tc thing in a process proc

**Comment by Kierany9** [comment]
It is done

File: code/game/gamemodes/traitor/king_disk.dm [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+		if(istype(disk_loc2, /mob))
+			var/mob/M = disk_loc2
+			var/datum/mind/traitor = M.mind
+			if(traitor.special_role)
+				if(N.king_timer >= 60)
+					N.king_timer = 0
+					var/list/all_items = traitor.current.GetAllContents()
+					if(traitor.special_role == "Mindslave")
+						for(var/datum/objective/protect/P in traitor.objectives)
+							all_items = P.target.current.GetAllContents()
```
Comment (Carbonhell) [inline code/game/gamemodes/traitor/king_disk.dm]: you have to check that the protect target isn't dead/gibbed or this will runtime

Comment (Kierany9) [inline code/game/gamemodes/traitor/king_disk.dm]: done

**Comment by Carbonhell** [comment]
code is k but nalar is needed for configs

**Comment by GeneralGoose** [comment]
we wuz kangz

**Comment by Spacedong** [comment]
I LIKE

File: code/game/gamemodes/traitor/king_disk.dm [hunk 13]
```diff
@@ -35,3 +35,25 @@
 	escape_objective.owner = traitor
 	traitor.objectives += escape_objective
 	return
+
+/datum/game_mode/traitor/king_disk/process()
+	for(var/obj/item/weapon/disk/nuclear/N in poi_list)
+		var/atom/disk_loc2 = N.loc
+		if(istype(disk_loc2, /mob))
+			var/mob/M = disk_loc2
+			var/datum/mind/traitor = M.mind
```
Comment (Chnkr) [inline code/game/gamemodes/traitor/king_disk.dm]: You might want to make sure the mob has a mind first, or you will run into some nasty runtimes.

Comment (Carbonhell) [inline code/game/gamemodes/traitor/king_disk.dm]: woops,i missed that,yep, definitely needed.

Comment (Kierany9) [inline code/game/gamemodes/traitor/king_disk.dm]: fixed
